Pokedex/CollectionViewController.swift

```diff
@@ -0,0 +1,183 @@
+        let layout = UICollectionViewFlowLayout()
+        layout.minimumLineSpacing = 0
+        layout.minimumInteritemSpacing = 0
+        collectionView = UICollectionView(frame: CGRect(x: 0, y: 124, width: self.view.frame.width, height: self.view.frame.height - 28) , collectionViewLayout: layout)
+        collectionView.register(PokemonCollectionViewCell.self, forCellWithReuseIdentifier: "pokeCell")
+        collectionView.backgroundColor = UIColor.white
+        collectionView.delegate = self
+        collectionView.dataSource = self
+        view.addSubview(collectionView)
+    }
+    
+    func setupTableView() {
+        tableView = UITableView(frame: CGRect(x: 0, y: 124, width: view.frame.width, height: view.frame.height - 28))
+        tableView.delegate = self
+        tableView.dataSource = self
+        tableView.register(PokemonTableViewCell.self, forCellReuseIdentifier: "pokeTableViewCell")
+        view.addSubview(tableView)
+    }
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return searchedPokemon.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "pokeTableViewCell", for: indexPath) as! PokemonTableViewCell
+        for subview in cell.contentView.subviews {
+            subview.removeFromSuperview()
+        }
+        cell.awakeFromNib()
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        let cell = cell as! PokemonTableViewCell
+        try! cell.pokemonImage.image = UIImage(data: Data(contentsOf: URL(string: searchedPokemon[indexPath.row].imageUrl)!))
+        cell.name.text = searchedPokemon[indexPath.row].name
+        cell.number.text = "# " + String(searchedPokemon[indexPath.row].number)
+    }
+    
+    func switchView(sender: UISegmentedControl) {
+        for subview in view.subviews {
+            subview.removeFromSuperview()
+        }
+        if (sender.selectedSegmentIndex == 0) {
+            view.addSubview(segmentedView)
+            setupFavoritesButton()
+            setupCollectionView()
+        } else {
+            view.addSubview(segmentedView)
+            setupFavoritesButton()
+            setupTableView()
+        }
+    }
+    
+    func viewFavoritesPage() {
+        performSegue(withIdentifier: "toFavorites", sender: self)
+    }
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == "toDetails" {
+            let next = segue.destination as! PokemonDetailsViewController
+            next.pokemon = selectedPokemon
+            next.favorites = defaults.object(forKey: "p") as! [Pokemon]!
+            next.defaults = defaults
+        } else {
+            let next = segue.destination as! FavoritesViewController
```

Comment:
need to ensure pokemon is not nil before segue to this view controller (or check for nil in the view controller)